compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -488,10 +494,36 @@ impl<'a> Conflicts<'a> {
     /// and must not be merged.
     fn record_statement_conflicts(&mut self, stmt: &Statement<'_>) {
         match &stmt.kind {
-            // While the left and right sides of an assignment must not overlap, we do not mark
-            // conflicts here as that would make this optimization useless. When we optimize, we
-            // eliminate the resulting self-assignments automatically.
-            StatementKind::Assign(_) => {}
+            StatementKind::Assign(v) => {
+                let p = v.0;
+                let rvalue = &v.1;
+                if p.is_indirect() {
+                    return;
+                }
+                let lhs_local = p.local;
+
+                struct RvalueVisitor<'a, 'b>(&'a mut Conflicts<'b>, Local);
+
+                impl<'a, 'b, 'tcx> Visitor<'tcx> for RvalueVisitor<'a, 'b> {
+                    fn visit_local(&mut self, local: &Local, _: PlaceContext, _: Location) {
+                        self.0.record_local_conflict(self.1, *local, "intra-assignment");
+                    }
+                }
+
+                let mut v = RvalueVisitor(self, lhs_local);
+                let location = Location { block: BasicBlock::MAX, statement_index: 0 };
+                if let Rvalue::Use(op) = rvalue {
+                    if let Some(rhs_place) = op.place() {
```

Comment:
Please leave a comment that this explicitly skips the place and only looks for index places in the projection